### ISSUE TITLE
[Feature] Add 2FA to Admin Authentication

### ISSUE THREAD

**Issue by gondolyr** [owncast/owncast#3695]
It would be a good security addition to add 2FA as an option, particularly [WebAuthn](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API), so that we can use security keys such as the [YubiKey](https://www.yubico.com/) or [Nitrokey](https://www.nitrokey.com/). It would help mitigate attack methods, such as brute-force, from attackers accessing the admin settings. As far as I can tell, the username cannot be changed from "admin" so an attacker would probably only need to start guessing the password.

Resources:
- https://fidoalliance.org/fido2-2/fido2-web-authentication-webauthn/

**Comment by gabek** [comment]
This is unlikely to be anything that I work on in the near future, but I'll keep this open in case somebody would be interested in looking into it.

**Comment by titanventura** [comment]
i'd like to pick this one up. please assign !

**Comment by gabek** [comment]
@titanventura Cool! This is a pretty big change. What do you have in mind? Right now we use HTTP Basic, and we can't get rid of that. How do you anticipate 2FA being added?

**Comment by titanventura** [comment]
Hi @gabek , i was thinking of adding a section to the server config tab of the server setup page in the admin dashboard. The admin can initially login using the default admin username and password, then setup 2FA using an `add OTP-less 2FA` (probably use a better name like `Configure 2FA` or something) button for registering their 2FA credentials.

the login part is where i'm a bit stuck. Can you help me understand how the current login form for admins work ? Is it something that comes out of the box by openapi ?

If it is a custom made alert form, i believe we should be able to prompt for passkeys after the user enters their username.

**Comment by gabek** [comment]
Currently, it uses `HTTP Basic` authentication. So that modal is presented by the browser, and we don't have any control over it. That's why I was concerned above about how it would work with our existing HTTP Basic auth. I guess it could be treated essentially as a dual layered auth, where after you pass HTTP Basic, the middleware checks if you also have passed 2FA, and if you haven't, it routes you to a new page where you perform 2FA. We'd need to introduce some kind of session token or something for the auth middleware to be aware of, so it knows if you've passed 2FA or not.

OpenAPI is just a specification defining the API, it doesn't provide any logic or implementation. 

For some additional clarification, when you say 2FA what exactly do you mean? Since it can mean a lot of things. A text message to a phone, using an authenticator app, hardware keys, and you mentioned passkeys (I actually don't know anything about passkeys).

**Comment by titanventura** [comment]
Yes, I think we could render the 2FA screen after basic auth by adding checks to the auth middleware. By 2FA, I meant WebAuthn (password-less, OTP-less auth like Fingerprint for MacOS, FaceID for IOS, Biometrics for android, passkeys like Yubikey etc.) as mentioned by @gondolyr.

The only thing I'm puzzled about is how the redirection in the frontend is going to happen after basic auth.

**Comment by gabek** [comment]
I think how it could work is that the auth middleware is fired when a request is made, and it sees that 2FA hasn't been completed, so it doesn't return the requested content, and instead returns the content for the 2FA challenge. So in that case there wouldn't be any redirection.
